apps/emqx_durable_storage/src/emqx_ds_storage_bitfield_lts.erl

```diff
@@ -0,0 +1,382 @@
+    rocksdb:db_handle(),
+    emqx_ds_storage_layer:gen_id(),
+    options()
+) ->
+    {schema(), emqx_ds_storage_layer:cf_refs()}.
+create(_ShardId, DBHandle, GenId, Options) ->
+    %% Get options:
+    BitsPerTopicLevel = maps:get(bits_per_wildcard_level, Options, 64),
+    TopicIndexBytes = maps:get(topic_index_bytes, Options, 4),
+    %% 10 bits -> 1024 ms -> ~1 sec
+    TSOffsetBits = maps:get(epoch_bits, Options, 10),
+    %% Create column families:
+    DataCFName = data_cf(GenId),
+    TrieCFName = trie_cf(GenId),
+    {ok, DataCFHandle} = rocksdb:create_column_family(DBHandle, DataCFName, []),
+    {ok, TrieCFHandle} = rocksdb:create_column_family(DBHandle, TrieCFName, []),
+    %% Create schema:
+    Schema = #{
+        bits_per_wildcard_level => BitsPerTopicLevel,
+        topic_index_bytes => TopicIndexBytes,
+        ts_bits => 64,
+        ts_offset_bits => TSOffsetBits
+    },
+    {Schema, [{DataCFName, DataCFHandle}, {TrieCFName, TrieCFHandle}]}.
+
+-spec open(
+    emqx_ds_replication_layer:shard_id(),
+    rocksdb:db_handle(),
+    emqx_ds_storage_layer:gen_id(),
+    emqx_ds_storage_layer:cf_refs(),
+    schema()
+) ->
+    s().
+open(_Shard, DBHandle, GenId, CFRefs, Schema) ->
+    #{
+        bits_per_wildcard_level := BitsPerTopicLevel,
+        topic_index_bytes := TopicIndexBytes,
+        ts_bits := TSBits,
+        ts_offset_bits := TSOffsetBits
+    } = Schema,
+    {_, DataCF} = lists:keyfind(data_cf(GenId), 1, CFRefs),
+    {_, TrieCF} = lists:keyfind(trie_cf(GenId), 1, CFRefs),
+    Trie = restore_trie(TopicIndexBytes, DBHandle, TrieCF),
+    %% If user's topics have more than learned 10 wildcard levels
+    %% (more than 2, really), then it's total carnage; learned topic
+    %% structure won't help.
+    MaxWildcardLevels = 10,
+    KeymapperCache = array:from_list(
+        [
+            make_keymapper(TopicIndexBytes, BitsPerTopicLevel, TSBits, TSOffsetBits, N)
+         || N <- lists:seq(0, MaxWildcardLevels)
+        ]
+    ),
+    #s{db = DBHandle, data = DataCF, trie = Trie, keymappers = KeymapperCache}.
+
+-spec store_batch(
+    emqx_ds_replication_layer:shard_id(), s(), [emqx_types:message()], emqx_ds:message_store_opts()
+) ->
+    emqx_ds:store_batch_result().
+store_batch(_ShardId, S = #s{db = DB, data = Data}, Messages, _Options) ->
+    lists:foreach(
+        fun(Msg) ->
+            {Key, _} = make_key(S, Msg),
+            Val = serialize(Msg),
+            rocksdb:put(DB, Data, Key, Val, [])
+        end,
+        Messages
+    ).
+
+get_streams(_Shard, #s{trie = Trie}, TopicFilter, _StartTime) ->
+    Indexes = emqx_ds_lts:match_topics(Trie, TopicFilter),
+    [#stream{storage_key = I} || I <- Indexes].
+
+make_iterator(_Shard, _Data, #stream{storage_key = StorageKey}, TopicFilter, StartTime) ->
+    %% Note: it's a good idea to keep the iterator structure lean,
+    %% since it can be stored on a remote node that could update its
+    %% code independently from us.
+    {ok, #it{
```

Comment:
General remark about those callbacks that are also RPC targets: maybe we should avoid using records for those, so we may avoid falling into the potential record change/evolution issues.  Even more so those that may be persisted?

Reply:
Persisting maps is not very efficient: every atom gets encoded separately. For example, `#{foo => #{foo => bar}}` becomes

```
0000000 203   t  \0  \0  \0 001   d  \0 003   f   o   o   t  \0  \0  \0
0000010 001   d  \0 003   f   o   o   d  \0 003   b   a   r            
000001d
```
If we strive to maintain scalability, I guess we'll have to embrace the pain of records.

Reply:
A Thrift-inspired approach is also worth considering: store `#{?FOO => #{?FOO => bar}}` where keys (e.g. `?FOO`) are short integer constants.

Reply:
Interesting approach indeed.

Reply:
https://emqx.atlassian.net/browse/EMQX-11279